src/wayland/shell/legacy/decoration.rs

```diff
@@ -0,0 +1,128 @@
+//!         KdeDecorationRequest::NewToplevelDecoration { surface } => (),
+//!         KdeDecorationRequest::RequestMode { mode, .. } => decoration.mode(mode),
+//!         KdeDecorationRequest::Release => (),
+//!     },
+//!     None,
+//! );
+
+use std::{cell::RefCell, ops::Deref, rc::Rc};
+
+use wayland_protocols::misc::server_decoration::server::{
+    org_kde_kwin_server_decoration::{Mode, OrgKdeKwinServerDecoration, Request as ServerDecorationRequest},
+    org_kde_kwin_server_decoration_manager::{
+        OrgKdeKwinServerDecorationManager, Request as ServerDecorationManagerRequest,
+    },
+};
+use wayland_server::protocol::wl_surface::WlSurface;
+use wayland_server::{Display, Filter, Global, Main};
+
+/// Events generated by the kde decoration manager.
+#[derive(Debug)]
+pub enum KdeDecorationRequest {
+    /// Decoration manager setup completion.
+    Setup {
+        /// KDE decoration manager.
+        manager: OrgKdeKwinServerDecorationManager,
+    },
+    /// A new toplevel decoration was instantiated.
+    NewToplevelDecoration {
+        /// KDE decoration object.
+        decoration: OrgKdeKwinServerDecoration,
+        /// The surface asosiated with the decoration.
+        surface: WlSurface,
+    },
+    /// Informs the compositor that the client prefers the provided decoration mode.
+    RequestMode {
+        /// KDE decoration object.
+        decoration: OrgKdeKwinServerDecoration,
+        /// The surface asosiated with the decoration.
+        surface: WlSurface,
+        /// Desired mode.
+        mode: Mode,
+    },
+    /// Decoration object was released.
+    Release {
+        /// The surface asosiated with the decoration.
+        surface: WlSurface,
+    },
+}
+
+/// Create a new KDE Decoration Manager global
+///
+/// A compositor can set the default mode for new clients at any time using
+/// [`OrgKdeKwinServerDecorationManager::default_mode`]. For individual windows the decorations can
+/// be set using [`OrgKdeKwinServerDecoration::mode`]. Both of these requests will cause the client
+/// to respond with its desired mode using [`KdeDecorationRequest::RequestMode`], which then should
+/// be acknowledged by the compositor using another [`OrgKdeKwinServerDecoration::mode`] event.
+/// **It is up to the compositor to prevent feedback loops**, a client is free to ignore the
+/// suggested decoration mode.
+pub fn init_kde_decoration_manager<L, Impl>(
+    display: &mut Display,
+    implementation: Impl,
+    _logger: L,
+) -> Global<OrgKdeKwinServerDecorationManager>
+where
+    L: Into<Option<slog::Logger>>,
+    Impl: FnMut(KdeDecorationRequest) + 'static,
+{
+    let callback = Rc::new(RefCell::new(implementation));
+    display.create_global(
+        1,
+        Filter::new(
+            move |(decoration_manager, _): (Main<OrgKdeKwinServerDecorationManager>, _), _, _| {
+                (&mut *callback.borrow_mut())(KdeDecorationRequest::Setup {
```

Comment:
I've noticed that when the interface is bound, the `org_kde_kwin_server_decoration_manager::default_mode` event is never sent to clients. The specificiation for the kde decorations says the following:

> This event is emitted directly after binding the interface. It contains the default mode for the decoration. When a new server decoration object is created this new object will be in the default mode until the first request_mode is requested.

Reply:
That's a good point, I've never considered downstream just not setting this in response to the setup event. I've played around with just passing `default_mode` too but I wanted to be able to override it in the future. Both could be done, but maybe it's just easier to always set the default mode at startup and scap the Setup part, since this is the old protocol anyway?

Reply:
So the "best" approach would be the most versatile:
- Provide a default-mode at `init`
- But return a handle that allows it too be changed later.

I am not sure, if this is worth for such an old protocol, but regarding @i509VCB 's comment, it should always be as hard as possible for downstream to violate/miss-use a protocol as sensibly possible. So sending the default_mode automatically at setup should be a thing.